File: 1061-lexicographically-smallest-equivalent-string/1061-lexicographically-smallest-equivalent-string.py
class Solution:
  def smallestEquivalentString(self, a: str, b: str, s: str) -> str:
    parent, ans = {}, ''
    for l, r in zip(a, b):
      if l in parent:
        while parent[l] != l:
          l = parent[parent[l]]
      else:
        parent[l] = l
      if r in parent:
        while parent[r] != r:
          r = parent[parent[r]]
      else:
        parent[r] = r
      if l > r:
        l, r = r, l
      parent[r] = l
    for k in s:
      if k in parent:
        while parent[k] != k:
          k = parent[parent[k]]
      ans += k
    return ans

        

File: 1061-lexicographically-smallest-equivalent-string/1061-lexicographically-smallest-equivalent-string.go
func smallestEquivalentString(a string, b string, s string) string {
  parent, ans := map[byte]byte{}, make([]byte, len(s))
  for i := range len(a) {
    l, r := a[i], b[i]
    if _, ok := parent[l]; ok {
      for parent[l] != l {
        l = parent[parent[l]]
      }
    } else {
      parent[l] = l
    }
    if _, ok := parent[r]; ok {
      for parent[r] != r {
        r = parent[parent[r]]
      }
    } else {
      parent[r] = r
    }
    if l > r {
      l, r = r, l
    }
    parent[r] = l
  }
  for i := range len(s) {
    k := s[i]
    if _, ok := parent[k]; ok {
      for parent[k] != k {
        k = parent[parent[k]]
      }
    }
    ans[i] = k
  }
  return string(ans)
}